exporter/otlpexporter/internal/arrow/stream.go

```diff
@@ -0,0 +1,246 @@
+// with the same index as the original status, for correlation.  Nil
+// channels will be returned when there are errors locating the
+// sender channel.
+func (s *Stream) getSenderChannels(statuses []*arrowpb.StatusMessage) ([]chan error, error) {
+	var err error
+
+	fin := make([]chan error, len(statuses))
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for idx, status := range statuses {
+		ch, ok := s.waiters[status.BatchId]
+		if !ok {
+			// Will break the stream.
+			err = multierr.Append(err, fmt.Errorf("duplicate stream response: %s", status.BatchId))
+			continue
+		}
+		delete(s.waiters, status.BatchId)
+		fin[idx] = ch
+	}
+
+	return fin, err
+}
+
+// processBatchStatus processes a single response from the server and unblocks the
+// associated senders.
+func (s *Stream) processBatchStatus(statuses []*arrowpb.StatusMessage) error {
+	fin, ret := s.getSenderChannels(statuses)
+
+	for idx, ch := range fin {
+		if ch == nil {
+			// In case getSenderChannels encounters a problem, the
+			// channel is nil.
+			continue
+		}
+		status := statuses[idx]
+
+		if status.StatusCode == arrowpb.StatusCode_OK {
+			ch <- nil
+			continue
+		}
+		var err error
+		switch status.ErrorCode {
+		case arrowpb.ErrorCode_UNAVAILABLE:
+			// TODO: translate retry information into the form
+			// exporterhelper recognizes.
+			err = fmt.Errorf("destination unavailable: %s: %s", status.BatchId, status.ErrorMessage)
+		case arrowpb.ErrorCode_INVALID_ARGUMENT:
+			err = consumererror.NewPermanent(
+				fmt.Errorf("invalid argument: %s: %s", status.BatchId, status.ErrorMessage))
+		default:
+			base := fmt.Errorf("unexpected stream response: %s: %s", status.BatchId, status.ErrorMessage)
+			err = consumererror.NewPermanent(base)
+
+			// Will break the stream.
+			ret = multierr.Append(ret, base)
+		}
+		ch <- err
+	}
+	return ret
+}
+
+// SendAndWait submits a batch of records to be encoded and sent.  Meanwhile, this
+// goroutine waits on the incoming context or for the asynchronous response to be
+// received by the stream reader.
+func (s *Stream) SendAndWait(ctx context.Context, records interface{}) error {
+	errCh := make(chan error, 1)
+	s.toWrite <- writeItem{
+		records: records,
+		errCh:   errCh,
+	}
+
+	// Note this ensures the caller's timeout is respected.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errCh:
+		return err
+	}
+}
+
+// encode produces the next batch of Arrow records.
+func (s *Stream) encode(records interface{}) (*arrowpb.BatchArrowRecords, error) {
+	// Note!! This is a placeholder.  After this PR merges the
+	// code base will be upgraded to the latest version of
+	// github.com/f5/otel-arrow-adapter, which returns one
+	// BatchArrowRecords per pdata Logs/Metrics/Traces.  The TODO below
+	// will be addressed, i.e., we will not drop all but one
+	// batch.
+	var batch *arrowpb.BatchArrowRecords
+	var err error
+	switch data := records.(type) {
```

Comment:
this is a nit, but we should probably have a default case here too that will error. It will most likely not be called/used ever, but it's good coverage.

Reply:
Fixed.